include/sys/dmu_objset.h

```diff
@@ -125,6 +126,12 @@ struct objset {
 	kmutex_t os_user_ptr_lock;
 	void *os_user_ptr;
 	sa_os_t *os_sa;
+
+	/* kernel thread to upgrade this dataset */
+	kthread_t *os_upgrade_thread;
+	kmutex_t os_upgrade_lock;
+	kcondvar_t os_upgrade_cv;
+	boolean_t os_upgrade_exit;
```

Comment:
That's certainly an option, although I don't think it's unreasonable to tie it to the spa.  My reasoning is that the major limiting factor should be the performance of the disk, not some shared system resource like cpu or memory.  Assuming that's true then I'd think we'd want to scale the number of upgrade threads with the number of pools.

We could still have a module option to control the number of threads in the taskq.  Since the vast majority of systems only have a single pool for many installations this would be effectively a system tunable.

I don't see any problems doing it the way you suggest, but I'm not sure it's simpler.